Content.Client/GameObjects/EntitySystems/DragDropSystem.cs

```diff
@@ -0,0 +1,402 @@
+                        _draggedEntity = entity;
+                        _draggable = draggable;
+                        _mouseDownTime = 0;
+                        _state = DragState.MOUSEDOWN;
+                        _mouseDownScreenPos = _inputManager.MouseScreenPosition;
+                        // don't want anything else to process the click,
+                        // but we will save the event so we can "re-play" it if this drag does
+                        // not turn into an actual drag so the click can be handled normally
+                        _savedMouseDown = args;
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private bool OnUseMouseUp(PointerInputCmdHandler.PointerInputCmdArgs args)
+        {
+            if (_state == DragState.MOUSEDOWN)
+            {
+                // quick mouseup, definitely treat it as a normal click by
+                // replaying the original
+                CancelDrag(true);
+                return false;
+            }
+            if (_state != DragState.DRAGGING) return false;
+
+            // remaining CancelDrag calls will not replay the click because
+            // by this time we've determined the input was actually a drag attempt
+
+
+            // tell the server we are dropping if we are over a valid drop target in range.
+            // We don't use args.EntityUid here because drag interactions generally should
+            // work even if there's something "on top" of the drop target
+            if (_interactionSystem.InRangeUnobstructed(_dragger.Transform.MapPosition,
+                    args.Coordinates.ToMap(_mapManager), ignoredEnt: _dragger) == false)
+            {
+                {
+                    CancelDrag(false);
+                    return false;
+                }
+            }
+
+            var entities = GameScreenBase.GetEntitiesUnderPosition(_stateManager, args.Coordinates);
+            foreach (var entity in entities)
+            {
+                // check if it's able to be dropped on by current dragged entity
+                if (_draggable.ClientCanDropOn(new CanDropEventArgs(_dragger, _draggedEntity, entity)))
+                {
+                    // tell the server about the drop attempt
+                    RaiseNetworkEvent(new DragDropMessage(args.Coordinates, _draggedEntity.Uid,
+                        entity.Uid));
+
+                    CancelDrag(false);
+                    return true;
+                }
+            }
+            CancelDrag(false);
+            return false;
+        }
+
+        private void StartDragging()
+        {
+            // this is checked elsewhere but adding this as a failsafe
+            if (_draggedEntity == null || _draggedEntity.Deleted)
+            {
+                Logger.Warning("Programming error. Cannot initiate drag, no dragged entity or entity" +
```

Comment:
If it's an error then this should probably be logged as, well, an error, no?
```suggestion
                Logger.Error("Programming error. Cannot initiate drag, no dragged entity or entity" +
```